src/System.IO.Pipelines.Networking.Tls/OpenSslSecurityContext.cs

```diff
@@ -0,0 +1,141 @@
+        internal IntPtr AplnBuffer => _alpnHandle.AddrOfPinnedObject();
+        internal int AplnBufferLength => _alpnSupportedProtocolsBuffer?.Length ?? 0;
+
+        private unsafe void SetupAlpn()
+        {
+            if (_alpnSupportedProtocols > 0)
+            {
+                //We need to get a buffer for the ALPN negotiation and pin it for sending to the lower API
+                _alpnSupportedProtocolsBuffer = ApplicationProtocols.GetBufferForProtocolId(_alpnSupportedProtocols, false);
+                _alpnHandle = GCHandle.Alloc(_alpnSupportedProtocolsBuffer, GCHandleType.Pinned);
+                Interop.SSL_CTX_set_alpn_protos(SslContext, AplnBuffer, (uint)AplnBufferLength);
+                if (_isServer)
+                {
+                    AlpnCallback = alpn_cb;
+                    Interop.SSL_CTX_set_alpn_select_cb(this);
+                }
+            }
+        }
+
+        private unsafe Interop.AlpnStatus alpn_cb(IntPtr ssl, out byte* selProto, out byte selProtoLen, byte* inProtos, int inProtosLen, IntPtr arg)
+        {
+            byte* matchBuffer = (byte*)_alpnHandle.AddrOfPinnedObject();
+            for (int i = 0; i < inProtosLen;)
+            {
+                var len = inProtos[i];
+                var s = new Span<byte>(inProtos + i + 1, len);
+                for (int x = 0; x < _alpnSupportedProtocolsBuffer.Length;)
+                {
+                    var len2 = matchBuffer[x];
+                    var sCompare = new Span<byte>(matchBuffer + x + 1, len2);
+                    if (sCompare.SequenceEqual(s))
+                    {
+                        selProto = matchBuffer + x + 1;
+                        selProtoLen = len2;
+                        return Interop.AlpnStatus.SSL_TLSEXT_ERR_OK;
+                    }
+                    x += len2 + 1;
+                }
+                i += len + 1;
+            }
+            selProto = null;
+            selProtoLen = 0;
+            return Interop.AlpnStatus.SSL_TLSEXT_ERR_NOACK;
+        }
+
+        private void SetupContext()
+        {
+            if (_isServer)
+            {
+                SslContext = Interop.NewServerContext(Interop.VerifyMode.SSL_VERIFY_NONE);
+            }
+            else
+            {
+                SslContext = Interop.NewClientContext(Interop.VerifyMode.SSL_VERIFY_NONE);
+            }
+
+            if (_certifcateInformation.Handle != IntPtr.Zero)
+            {
+                Interop.SetKeys(SslContext, _certifcateInformation.CertificateHandle, _certifcateInformation.PrivateKeyHandle);
+            }
+        }
+
+        public ISecurePipeline CreateSecurePipeline(IPipelineConnection pipeline)
+        {
+            var ssl = Interop.SSL_new(SslContext);
+            var chan = new SecurePipeline<OpenSslConnectionContext>(pipeline, _pipelineFactory, new OpenSslConnectionContext(this, ssl));
+            return chan;
+        }
+
+        public void Dispose()
```

Comment:
Finalize will fix it, safehandle == allocation, what does it bring to the party that over dispose/finalize?